File: src/services/teacher.service.ts
const API_URL = 'http://127.0.0.1:8000/api/teachers';

export interface Teacher {
  id: string;
  user_id?: string;
  school_id: string;
  teacher_id: string;
  employee_id: string;
  teacher_name: string;
  email: string;
  phone?: string;
  department?: string;
  designation?: string;
  qualification?: string;
  joining_date?: string;
  profile_photo_url?: string;
  status: 'active' | 'inactive' | 'pending';
  created_at?: string;
  updated_at?: string;
  schools?: {
    school_name: string;
    school_code: string;
  };
}

export interface TeacherStats {
  assignedStudents: number;
  completedAssessments: number;
  generatedReports: number;
  activeInterventions: number;
}

export const teacherService = {
  createTeacher: async (data: Partial<Teacher> & { temp_password?: string }): Promise<Teacher> => {
    const response = await fetch(API_URL, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(data)
    });
    const result = await response.json();
    if (!response.ok) throw new Error(result.detail || 'Failed to create teacher');
    return result;
  },

  getTeachers: async (): Promise<Teacher[]> => {
    const response = await fetch(API_URL);
    if (!response.ok) throw new Error('Failed to fetch teachers');
    return await response.json();
  },

  getTeacherById: async (id: string): Promise<Teacher> => {
    const response = await fetch(`${API_URL}/${id}`);
    if (!response.ok) throw new Error('Failed to fetch teacher details');
    return await response.json();
  },

  getTeachersBySchool: async (schoolId: string): Promise<Teacher[]> => {
    const response = await fetch(`${API_URL}/school/${schoolId}`);
    if (!response.ok) throw new Error('Failed to fetch teachers for school');
    return await response.json();
  },

  updateTeacher: async (id: string, data: Partial<Teacher> & { temp_password?: string }): Promise<Teacher> => {
    const response = await fetch(`${API_URL}/${id}`, {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(data)
    });
    const result = await response.json();
    if (!response.ok) throw new Error(result.detail || 'Failed to update teacher');
    return result;
  },

  updateTeacherStatus: async (id: string, status: Teacher['status']): Promise<Teacher> => {
    const response = await fetch(`${API_URL}/${id}/status`, {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ status })
    });
    const result = await response.json();
    if (!response.ok) throw new Error(result.detail || 'Failed to update teacher status');
    return result;
  },

  deactivateTeacher: async (id: string): Promise<Teacher> => {
    return teacherService.updateTeacherStatus(id, 'inactive');
  },

  getTeacherStats: async (id: string): Promise<TeacherStats> => {
    const response = await fetch(`${API_URL}/${id}/stats`);
    if (!response.ok) throw new Error('Failed to fetch teacher stats');
    return await response.json();
  }
};

File: backend/services/teacher_service.py
from typing import List, Dict, Any, Optional
from services.supabase_client import get_supabase_client
from uuid import UUID

class TeacherService:
    @staticmethod
    def create_teacher(data: Dict[str, Any]) -> Dict[str, Any]:
        """Create a new teacher"""
        try:
            supabase = get_supabase_client()
            # Map standard fields
            teacher_data = {
                "school_id": data.get("school_id"),
                "teacher_id": data.get("teacher_id"),
                "employee_id": data.get("employee_id"),
                "teacher_name": data.get("teacher_name"),
                "email": data.get("email"),
                "phone": data.get("phone"),
                "department": data.get("department"),
                "designation": data.get("designation"),
                "qualification": data.get("qualification"),
                "joining_date": data.get("joining_date"),
                "temp_password": data.get("temp_password"),
                "status": data.get("status", "active")
            }
            
            response = supabase.table('teacher_profiles').insert(teacher_data).execute()
            if not response.data:
                raise Exception("Failed to create teacher")
                
            return response.data[0]
        except Exception as e:
            print(f"Error creating teacher: {e}")
            raise Exception(f"Failed to create teacher: {str(e)}")

    @staticmethod
    def get_teachers() -> List[Dict[str, Any]]:
        """Get all teachers"""
        try:
            supabase = get_supabase_client()
            # Fetch teachers with basic school info
            response = supabase.table('teacher_profiles').select(
                '*, schools (school_name, school_code)'
            ).execute()
            
            return response.data
        except Exception as e:
            print(f"Error fetching teachers: {e}")
            return []

    @staticmethod
    def get_teacher_by_id(teacher_id: str) -> Optional[Dict[str, Any]]:
        """Get a teacher by ID"""
        try:
            supabase = get_supabase_client()
            response = supabase.table('teacher_profiles').select(
                '*, schools (school_name, school_code)'
            ).eq('id', teacher_id).execute()
            
            if response.data:
                return response.data[0]
            return None
        except Exception as e:
            print(f"Error fetching teacher: {e}")
            return None

    @staticmethod
    def get_teachers_by_school(school_id: str) -> List[Dict[str, Any]]:
        """Get all teachers assigned to a specific school"""
        try:
            supabase = get_supabase_client()
            response = supabase.table('teacher_profiles').select(
                '*'
            ).eq('school_id', school_id).execute()
            return response.data
        except Exception as e:
            print(f"Error fetching teachers by school: {e}")
            return []

    @staticmethod
    def update_teacher(teacher_id: str, data: Dict[str, Any]) -> Dict[str, Any]:
        """Update a teacher"""
        try:
            supabase = get_supabase_client()
            response = supabase.table('teacher_profiles').update(data).eq('id', teacher_id).execute()
            if not response.data:
                raise Exception("Teacher not found or update failed")
            return response.data[0]
        except Exception as e:
            print(f"Error updating teacher: {e}")
            raise Exception(f"Failed to update teacher: {str(e)}")

    @staticmethod
    def deactivate_teacher(teacher_id: str) -> Dict[str, Any]:
        """Soft delete a teacher by setting status to inactive"""
        return TeacherService.update_teacher(teacher_id, {"status": "inactive"})

    @staticmethod
    def get_teacher_stats(teacher_id: str) -> Dict[str, int]:
        """Get stats for a teacher"""
        try:
            supabase = get_supabase_client()
            
            # Fetch assigned students count
            assignments_resp = supabase.table('teacher_student_assignments').select(
                'id', count='exact'
            ).eq('teacher_id', teacher_id).eq('status', 'active').execute()
            
            assigned_students_count = assignments_resp.count if assignments_resp.count is not None else 0
            
            # Placeholders for future phases
            return {
                "assignedStudents": assigned_students_count,
                "completedAssessments": 0,
                "generatedReports": 0,
                "activeInterventions": 0
            }
        except Exception as e:
            print(f"Error fetching teacher stats: {e}")
            return {
                "assignedStudents": 0,
                "completedAssessments": 0,
                "generatedReports": 0,
                "activeInterventions": 0
            }

teacher_service = TeacherService()
